File: sdks/swift/Sources/CheeTaxi/Resources.swift
import Foundation
import Crypto

// MARK: - Models

public struct GeoPoint: Codable {
    public let lat: Double
    public let lng: Double
    public let address: String?
    public init(lat: Double, lng: Double, address: String? = nil) {
        self.lat = lat; self.lng = lng; self.address = address
    }
}

public struct Trip: Codable {
    public let id: String
    public let publicId: String
    public let status: String
    public let mode: String
    public let totalFare: Double
    public let currency: String
}

public struct FareQuote: Codable {
    public let baseFare: Double
    public let distanceFare: Double
    public let totalFare: Double
    public let currency: String
    public let distanceMeters: Double
    public let durationSeconds: Int
}

// MARK: - Resources

public struct TripsResource {
    let config: CheeTaxiConfig

    public func request(
        pickup: GeoPoint, dropoff: GeoPoint,
        mode: String, vehicleType: String, paymentMethod: String,
        promoCode: String? = nil
    ) async throws -> [String: Any] {
        let body: [String: Any] = [
            "pickup": ["lat": pickup.lat, "lng": pickup.lng, "address": pickup.address as Any],
            "dropoff": ["lat": dropoff.lat, "lng": dropoff.lng, "address": dropoff.address as Any],
            "mode": mode, "vehicleType": vehicleType, "paymentMethod": paymentMethod,
        ]
        return try await cheetaxiRequest(config, method: "POST", path: "/trips/request", body: body) as? [String: Any] ?? [:]
    }

    public func get(_ tripId: String) async throws -> [String: Any] {
        return try await cheetaxiRequest(config, method: "GET", path: "/trips/\(tripId)") as? [String: Any] ?? [:]
    }

    public func cancel(_ tripId: String, reason: String) async throws -> [String: Any] {
        return try await cheetaxiRequest(config, method: "POST", path: "/trips/\(tripId)/cancel", body: ["reason": reason]) as? [String: Any] ?? [:]
    }
}

public struct PricingResource {
    let config: CheeTaxiConfig

    public func quote(
        vehicleType: String,
        pickupLat: Double, pickupLng: Double,
        dropoffLat: Double, dropoffLng: Double
    ) async throws -> [String: Any] {
        let qs = "vehicleType=\(vehicleType)&pickupLat=\(pickupLat)&pickupLng=\(pickupLng)&dropoffLat=\(dropoffLat)&dropoffLng=\(dropoffLng)"
        return try await cheetaxiRequest(config, method: "GET", path: "/pricing/quote?\(qs)") as? [String: Any] ?? [:]
    }

    public func listTiers() async throws -> [Any] {
        return try await cheetaxiRequest(config, method: "GET", path: "/pricing/tiers") as? [Any] ?? []
    }
}

public struct SubscriptionsResource {
    let config: CheeTaxiConfig
    public func listPlans() async throws -> [Any] {
        return try await cheetaxiRequest(config, method: "GET", path: "/subscriptions/plans") as? [Any] ?? []
    }
    public func getMyActive() async throws -> [String: Any] {
        return try await cheetaxiRequest(config, method: "GET", path: "/subscriptions/me/active") as? [String: Any] ?? [:]
    }
}

public struct WalletsResource {
    let config: CheeTaxiConfig
    public func getMyWallet() async throws -> [String: Any] {
        return try await cheetaxiRequest(config, method: "GET", path: "/wallets/me") as? [String: Any] ?? [:]
    }
}

public struct WebhooksResource {
    public init() {}

    /// Verify a webhook signature using HMAC-SHA256.
    public func verifySignature(body: String, signature: String, secret: String) -> Bool {
        let key = SymmetricKey(data: Data(secret.utf8))
        let hmac = HMAC<SHA256>.authenticationCode(for: Data(body.utf8), using: key)
        let computed = Data(hmac).map { String(format: "%02x", $0) }.joined()
        return computed == signature
    }
}

public struct HealthResource {
    let config: CheeTaxiConfig
    public func liveness() async throws -> [String: Any] {
        return try await cheetaxiRequest(config, method: "GET", path: "/health") as? [String: Any] ?? [:]
    }
    public func readiness() async throws -> [String: Any] {
        return try await cheetaxiRequest(config, method: "GET", path: "/health/ready") as? [String: Any] ?? [:]
    }
}

File: sdks/kotlin/cheetaxi/src/main/kotlin/com/cheetaxi/sdk/Resources.kt
package com.cheetaxi.sdk

import org.json.JSONArray
import org.json.JSONObject
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

data class GeoPoint(val lat: Double, val lng: Double, val address: String? = null) {
    fun toJson() = JSONObject().apply {
        put("lat", lat); put("lng", lng); address?.let { put("address", it) }
    }
}

class TripsResource(private val config: CheeTaxiConfig) {
    suspend fun request(
        pickup: GeoPoint, dropoff: GeoPoint,
        mode: String, vehicleType: String, paymentMethod: String,
        promoCode: String? = null
    ): JSONObject {
        val body = JSONObject().apply {
            put("pickup", pickup.toJson())
            put("dropoff", dropoff.toJson())
            put("mode", mode)
            put("vehicleType", vehicleType)
            put("paymentMethod", paymentMethod)
            promoCode?.let { put("promoCode", it) }
        }
        return CheeTaxiClient(config).request("POST", "/trips/request", body) as JSONObject
    }

    suspend fun get(tripId: String): JSONObject =
        CheeTaxiClient(config).request("GET", "/trips/$tripId") as JSONObject

    suspend fun cancel(tripId: String, reason: String, by: String = "passenger"): JSONObject {
        val body = JSONObject().apply { put("reason", reason); put("by", by) }
        return CheeTaxiClient(config).request("POST", "/trips/$tripId/cancel", body) as JSONObject
    }

    suspend fun accept(tripId: String): JSONObject =
        CheeTaxiClient(config).request("POST", "/trips/$tripId/accept") as JSONObject

    suspend fun arrive(tripId: String): JSONObject =
        CheeTaxiClient(config).request("POST", "/trips/$tripId/arrive") as JSONObject

    suspend fun start(tripId: String): JSONObject =
        CheeTaxiClient(config).request("POST", "/trips/$tripId/start") as JSONObject

    suspend fun complete(tripId: String, actualDistanceMeters: Double? = null, actualDurationSeconds: Int? = null): JSONObject {
        val body = JSONObject()
        actualDistanceMeters?.let { body.put("actualDistanceMeters", it) }
        actualDurationSeconds?.let { body.put("actualDurationSeconds", it) }
        return CheeTaxiClient(config).request("POST", "/trips/$tripId/complete", body) as JSONObject
    }
}

class PricingResource(private val config: CheeTaxiConfig) {
    suspend fun quote(
        vehicleType: String,
        pickupLat: Double, pickupLng: Double,
        dropoffLat: Double, dropoffLng: Double
    ): JSONObject {
        val path = "/pricing/quote?vehicleType=$vehicleType&pickupLat=$pickupLat&pickupLng=$pickupLng&dropoffLat=$dropoffLat&dropoffLng=$dropoffLng"
        return CheeTaxiClient(config).request("GET", path) as JSONObject
    }

    suspend fun listTiers(): JSONArray =
        CheeTaxiClient(config).request("GET", "/pricing/tiers") as JSONArray
}

class SubscriptionsResource(private val config: CheeTaxiConfig) {
    suspend fun listPlans(): JSONArray =
        CheeTaxiClient(config).request("GET", "/subscriptions/plans") as JSONArray

    suspend fun getMyActive(): JSONObject =
        CheeTaxiClient(config).request("GET", "/subscriptions/me/active") as JSONObject
}

class WalletsResource(private val config: CheeTaxiConfig) {
    suspend fun getMyWallet(): JSONObject =
        CheeTaxiClient(config).request("GET", "/wallets/me") as JSONObject

    suspend fun topUp(amount: Double, currency: String, provider: String) {
        val body = JSONObject().apply {
            put("amount", amount); put("currency", currency); put("provider", provider)
        }
        CheeTaxiClient(config).request("POST", "/wallets/me/topup", body)
    }
}

class WebhooksResource {
    /// Verify a webhook signature using HMAC-SHA256.
    fun verifySignature(body: String, signature: String, secret: String): Boolean {
        val key = SecretKeySpec(secret.toByteArray(), "HmacSHA256")
        val mac = Mac.getInstance("HmacSHA256").apply { init(key) }
        val computed = mac.doFinal(body.toByteArray()).joinToString("") { "%02x".format(it) }
        return computed == signature
    }
}

class HealthResource(private val config: CheeTaxiConfig) {
    suspend fun liveness(): JSONObject =
        CheeTaxiClient(config).request("GET", "/health") as JSONObject

    suspend fun readiness(): JSONObject =
        CheeTaxiClient(config).request("GET", "/health/ready") as JSONObject
}
